lib/teleterm/daemon/daemon_headless_watcher.go

```diff
@@ -0,0 +1,182 @@
+	defer s.headlessWatcherClosersMu.Unlock()
+
+	cluster, err := s.ResolveCluster(uri)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = s.startHeadlessWatcher(cluster)
+	return trace.Wrap(err)
+}
+
+// StartHeadlessWatchers starts headless watchers for all connected clusters.
+func (s *Service) StartHeadlessWatchers() error {
+	s.headlessWatcherClosersMu.Lock()
+	defer s.headlessWatcherClosersMu.Unlock()
+
+	clusters, err := s.cfg.Storage.ReadAll()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for _, c := range clusters {
+		if c.Connected() {
+			if err := s.startHeadlessWatcher(c); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// startHeadlessWatcher starts a background process to watch for pending headless
+// authentications for this cluster.
+func (s *Service) startHeadlessWatcher(cluster *clusters.Cluster) error {
+	// If there is already a watcher for this cluster, close and replace it.
+	// This may occur after relogin, for example.
+	if err := s.stopHeadlessWatcher(cluster.URI.String()); err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+
+	retry, err := retryutils.NewLinear(retryutils.LinearConfig{
+		First:  utils.FullJitter(defaults.HighResPollingPeriod / 10),
+		Step:   defaults.HighResPollingPeriod / 5,
+		Max:    defaults.HighResPollingPeriod,
+		Jitter: retryutils.NewHalfJitter(),
+		Clock:  s.cfg.Storage.Clock,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	watchCtx, watchCancel := context.WithCancel(s.closeContext)
+	s.headlessWatcherClosers[cluster.URI.String()] = watchCancel
+
+	watch := func() error {
+		watcher, closeWatcher, err := cluster.WatchPendingHeadlessAuthentications(watchCtx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		retry.Reset()
+
+		defer closeWatcher()
+		for {
+			select {
+			case event := <-watcher.Events():
+				// Ignore non-put events.
+				if event.Type != types.OpPut {
+					continue
+				}
+
+				ha, ok := event.Resource.(*types.HeadlessAuthentication)
+				if !ok {
+					return trace.Errorf("headless watcher returned an unexpected resource type %T", event.Resource)
+				}
+
+				// Notify the Electron App of the pending headless authentication to handle resolution.
+				req := &api.SendPendingHeadlessAuthenticationRequest{
+					RootClusterUri:                 cluster.URI.String(),
+					HeadlessAuthenticationId:       ha.GetName(),
+					HeadlessAuthenticationClientIp: ha.ClientIpAddress,
+				}
+
+				if _, err := s.tshdEventsClient.SendPendingHeadlessAuthentication(watchCtx, req); err != nil {
+					return trace.Wrap(err)
+				}
+			case <-watcher.Done():
+				return trace.Wrap(err)
+			case <-watchCtx.Done():
+				return nil
+			}
+		}
+	}
+
+	log := s.cfg.Log.WithField("cluster", cluster.URI.String())
+	log.Debugf("Starting headless watch loop.")
+	go func() {
+		defer watchCancel()
+		for {
+			if !cluster.Connected() {
+				log.Debugf("Not connected to cluster. Returning from headless watch loop.")
+				return
```

Comment:
Assuming this happens when cluster is expired? Should it remove itself from `headlessWatcherClosers`